### ISSUE TITLE
Export skips some columns with user defined types

### ISSUE THREAD

**Issue by lupynos** [darold/ora2pg#614]
We have here another case when export skips some columns with user defined types.

Types:
```sql
CREATE OR REPLACE TYPE geomext AS OBJECT (
   x               DOUBLE PRECISION,
      -- x coordinate
   y               DOUBLE PRECISION,
      -- y coordinate
   rotation        DOUBLE PRECISION,
      -- element rotation    
   deg_start       DOUBLE PRECISION,
      -- arc/ellipse start angle (in radians)    
   deg_sweep       DOUBLE PRECISION,
      -- arc/ellipse sweep angle (in radians)    
   prim_axis       DOUBLE PRECISION,
      -- ellipse primary axis diameter    
   sec_axis        DOUBLE PRECISION,
      -- ellipse secondary axis diameter    
   width           DOUBLE PRECISION,
      -- element width    
   height          DOUBLE PRECISION,
      -- element height    
   justification   INTEGER,
      -- text justification    
   font            VARCHAR2 (255 CHAR),
      -- text font (deprecated)    
   text            VARCHAR2 (1000 CHAR)
      -- text   
);
/
CREATE OR REPLACE TYPE feature_link AS OBJECT (
   ftid_ref         VARCHAR2(50),  
      -- Link to a feature type
   fid_ref          NUMBER(38)  
      -- Link to a feature
);
/
CREATE OR REPLACE TYPE gtag_component AS OBJECT (
   type             NUMBER(6),          
      -- geometry type
   source_type      NUMBER(4),          
      -- geometry type in CAD
   meta_type_id     VARCHAR2(50 CHAR),  
      -- Component metadata id
   name             VARCHAR2(100 CHAR), 
      -- Optional name of a component
   extension_info   VARCHAR2(500 CHAR),  
      -- Optional information for extensions
   feature_ref      feature_link, 
      -- Optional information for extensions
   symbology_tokens VARCHAR2(2000 CHAR)
);
/
CREATE OR REPLACE TYPE geomext_array AS
  VARRAY (10000) OF geomext;
/
CREATE OR REPLACE TYPE gtag_component_array AS
      VARRAY (10000) OF gtag_component;
/
```

Table and inserts:
```sql
SET DEFINE OFF;

CREATE TABLE "TEST" 
   (	"ID" NUMBER(38,0), 
	"TYPE_ID" VARCHAR2(50 CHAR), 
	"FID" NUMBER(38,0), 
	"SOURCE" NUMBER(2,0), 
	"COMPLEX_GEOMEXT" "GEOMEXT_ARRAY" , 
	"EXTENSION_INFO" VARCHAR2(1000 CHAR), 
	"COMPLEX_INFO" "GTAG_COMPONENT_ARRAY" 
   );

  CREATE UNIQUE INDEX "PK_E__TEST_ID" ON "TEST" ("ID");

  CREATE INDEX "IDX_E__TEST_FID" ON "TEST" ("FID");

  ALTER TABLE "TEST" ADD CONSTRAINT "PK_E__TEST_ID" PRIMARY KEY ("ID")
  USING INDEX  ENABLE;

Insert into TEST ("ID",TYPE_ID,FID,"SOURCE",COMPLEX_GEOMEXT,EXTENSION_INFO,COMPLEX_INFO) values ('1709782591331958081341946','ft_e_route_gtcs_0002','77791610698075439828446366202','0',DMK.GEOMEXT_ARRAY(DMK.GEOMEXT(3451185.21878925, 5486129.96564963, -1.5707963267949, 0, 0, 0, 0, 0, 0, 7, NULL, NULL), DMK.GEOMEXT(3451185.21878925, 5486129.96564963, -1.5707963267949, 0, 0, 0, 0, 0, 0, 7, NULL, NULL), DMK.GEOMEXT(3451185.21878925, 5486129.96564963, -1.5707963267949, 0, 0, 0, 0, 0, 0, 7, NULL, NULL), DMK.GEOMEXT(3451185.21878925, 5486129.96564963, -1.5707963267949, 0, 0, 0, 0, 0, 0, 7, NULL, NULL), DMK.GEOMEXT(3451185.21878925, 5486129.96564963, -1.5707963267949, 0, 0, 0, 0, 0, 0, 7, NULL, NULL), DMK.GEOMEXT(3451185.21878925, 5486129.96564963, -1.5707963267949, 0, 0, 0, 0, 0, 0, 7, NULL, NULL), null, null, DMK.GEOMEXT(0, 0, -1.5707963267949, 0, 0, 0, 0, 0, 0, NULL, NULL, NULL)),'translate=3451185.21878925,5486129.96564963&rotation=-1.5707963267949"&scale="1.0,1.0',DMK.GTAG_COMPONENT_ARRAY(DMK.GTAG_COMPONENT(1000, 0, 'gtcs_0002_03', NULL, NULL, DMK.FEATURE_LINK('ft_e_cable_hv', 77791616435012846752116918778), NULL), DMK.GTAG_COMPONENT(1000, 0, 'gtcs_0002_02', NULL, NULL, DMK.FEATURE_LINK('ft_e_cable_lv', 77791616471906334899536022010), 'st_gtcs_0002_02_01(st_gtcs_0002_02_01_else)'), DMK.GTAG_COMPONENT(1000, 0, 'gtcs_0002_t_cable', NULL, NULL, DMK.FEATURE_LINK('ft_telco_cable', 77791615420441922698091579898), NULL), DMK.GTAG_COMPONENT(1000, 0, 'gtcs_0002_01', NULL, NULL, DMK.FEATURE_LINK('ft_e_protpipe', 77791610827202648344413227514), 'st_gtcs_0002_01_02(rule_gtcs_0002_01_02_color1,rule_gtcs_0002_01_02_else) st_gtcs_0002_01_01(rule_gtcs_0002_symb_else)'), DMK.GTAG_COMPONENT(1000, 0, 'gtcs_0002_01', NULL, NULL, DMK.FEATURE_LINK('ft_e_protpipe', 77791610790309160196994124282), 'st_gtcs_0002_01_02(rule_gtcs_0002_01_02_color_else,rule_gtcs_0002_01_02_else) st_gtcs_0002_01_01(rule_gtcs_0002_symb_else)'), DMK.GTAG_COMPONENT(1000, 0, 'gtcs_0002_01', NULL, NULL, DMK.FEATURE_LINK('ft_e_protpipe', 77791610753415672049575021050), 'st_gtcs_0002_01_02(rule_gtcs_0002_01_02_color_else,rule_gtcs_0002_01_02_else) st_gtcs_0002_01_01(rule_gtcs_0002_symb_else)'), DMK.GTAG_COMPONENT(1001, 0, 'gtcs_0002_06', NULL, NULL, NULL, NULL), DMK.GTAG_COMPONENT(1001, 0, 'gtcs_0002_09', NULL, NULL, NULL, NULL), DMK.GTAG_COMPONENT(1000, 0, 'gtcs_0002_08', NULL, NULL, NULL, NULL)));

Insert into TEST (ID,TYPE_ID,FID,SOURCE,COMPLEX_GEOMEXT,EXTENSION_INFO,COMPLEX_INFO) values ('12147323909052690307365','ft_e_route_gtda_5060002','7020093067627421913553189','0',DMK.GEOMEXT_ARRAY(DMK.GEOMEXT(0, 0, 0, 0, 0, 0, 0, 0, 0, NULL, NULL, NULL), DMK.GEOMEXT(0, 0, 0, 0, 0, 0, 0, 0, 0, NULL, NULL, NULL), DMK.GEOMEXT(0, 0, 0, 0, 0, 0, 0, 0, 0, NULL, NULL, NULL), DMK.GEOMEXT(0, 0, 0, 0, 0, 0, 0, 0, 0, NULL, NULL, NULL)),null,DMK.GTAG_COMPONENT_ARRAY(DMK.GTAG_COMPONENT(1001, 0, 'gtda_5060002_2', NULL, NULL, NULL, NULL), DMK.GTAG_COMPONENT(1001, 0, 'gtda_5060002_1', NULL, NULL, NULL, NULL), DMK.GTAG_COMPONENT(1001, 0, 'gtda_5060002_2', NULL, NULL, NULL, NULL), DMK.GTAG_COMPONENT(1001, 0, 'gtda_5060002_1', NULL, NULL, NULL, NULL)));
```

**Comment by darold** [comment]
Thanks for the sample to reproduce. After import in Oracle and export using Ora2Pg I have the following translation:
```
CREATE TYPE geomext AS (
   x               double precision,
   y               double precision,
   rotation        double precision,
   deg_start       double precision,
   deg_sweep       double precision,
   prim_axis       double precision,
   sec_axis        double precision,
   width           double precision,
   height          double precision,
   justification   numeric,
   font            varchar(255),
   text            varchar(1000)
);
CREATE TYPE geomext_array AS (geomext_array geomext[10000]);

CREATE TYPE feature_link AS (
   ftid_ref         varchar(50),
   fid_ref          bigint

);

CREATE TYPE gtag_component AS (
   type             numeric,
   source_type      smallint,
   meta_type_id     varchar(50),
   name             varchar(100),
   extension_info   varchar(500),
   feature_ref      feature_link,
   symbology_tokens varchar(2000)
);

CREATE TYPE gtag_component_array AS (gtag_component_array gtag_component[10000]);

CREATE TABLE test (
        id numeric(38) NOT NULL,
        type_id varchar(50),
        fid numeric(38),
        source smallint,
        complex_geomext GEOMEXT_ARRAY,
        extension_info varchar(1000),
        complex_info GTAG_COMPONENT_ARRAY
) ;
```
which is loaded without error into PostgreSQL. Until here everything is ok, the problem comes with data export. Using INSERT for better reading I got that:
```
INSERT INTO test (id,type_id,fid,source,complex_geomext,extension_info,complex_info) VALUES
 (1709782591331958081341946,E'ft_e_route_gtcs_0002',77791610698075439828446366202,0,
 ROW(),E'translate=3451185.21878925,5486129.96564963&rotation=-1.5707963267949"&scale="1.0,1.0',
 ROW()
);
INSERT INTO test (id,type_id,fid,source,complex_geomext,extension_info,complex_info) VALUES
  (12147323909052690307365,E'ft_e_route_gtda_5060002',7020093067627421913553189,0,
  ROW(),NULL,ROW());
```
Data in user defined type in this case are exported as empty `ROW()`. The problem is that Oracle or DBD::Oracle doesn't return anything for these columns even with `ora_objects` set to 1 on the database handle. I'm investigating the reason.

**Comment by lupynos** [comment]
Hi Darold, can we help somehow in this issue, prepare something or rather look for other solution?

**Comment by darold** [comment]
Hi,

I have found the source of the problem. Actually I use `fetchall_arrayref(undef, DATA_LIMIT)` to extract data in bulk of DATA_LIMIT rows but it seems that `fetchall_arrayref()` is not able to associate complex custom types to the returned array.

I have now an external test code that can reproduce the problem and I'm working to get a solution asap.

Regards,

**Comment by darold** [comment]
Just to let you know that I have something working but I have still an issue with the export of the numeric(38) from the nested feature_link type. Value `77791616435012846752116918778` is exported as `7.77916164350128e+28`, the use of bignum in Perl script doesn't help. I need to run some regression test too because there is a major change in data export. I hope to close the issue this weekend.

**Comment by darold** [comment]
Commit 5e166cc might fix this issue. Note that PostgreSQL support arrays natively so instead of using nested types like done in oracle it could be better to declare your table as follow:
```
CREATE TABLE test (
        id numeric(38) NOT NULL,
        type_id varchar(50),
        fid numeric(38),
        source smallint,
        complex_geomext GEOMEXT[],
        extension_info varchar(1000),
        complex_info GTAG_COMPONENT[]
) ;
```
Then you could remove type GEOMEXT_ARRAY and GTAG_COMPONENT_ARRAY. To be able to import data with theses changes you will have to set the following on your ora2pg.conf:
```
MODIFY_TYPE     TEST:COMPLEX_GEOMEXT:geomext[],TEST:COMPLEX_INFO:gtag_component[]
```
Lost of precision with big numbers in nested type feature_link is still under investigation.

**Comment by darold** [comment]
The feature_link type is declared as follow:
```
CREATE OR REPLACE TYPE feature_link AS OBJECT (
   ftid_ref         VARCHAR2(50),  
      -- Link to a feature type
   fid_ref          NUMBER(38)  
      -- Link to a feature
);
```
If I declare column `fid_ref` as a `varchar2(50)` like column `ftid_ref` data export is well done. This can be taken as a workaround for now, I will not be able to work again on it until next weekend.

**Comment by lupynos** [comment]
with latest master version I have problem with perl interpreter, it crashes (on windows strawberry perl - it could be problem) ... but I hope I will test it soon

**Comment by Bulva** [comment]
Hi Darold,

we have made some tests with Python and cx_Oracle related to exporting a user defined types and we found that there was a problem with the `gtag_component` as well. The problem was the version of [odpi](https://github.com/oracle/odpi) where was an error with a conversion of data types. When we update cx_Oracle package to the version which uses a new odpi export of `gtag_component` in Python works now.

I am not able to find if DBD::Oracle uses odpi but if yes, there can be the same problem. It is not solution of problem in this thread but it can be a clue for you.
